// node-red-contrib-nora/build/nodes/nora-blinds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var nora_1 = require("../nora");
var util_1 = require("./util");
module.exports = function (RED) {
    RED.nodes.registerType('nora-blinds', function (config) {
        var _this = this;
        RED.nodes.createNode(this, config);
        var noraConfig = RED.nodes.getNode(config.nora);
        if (!noraConfig || !noraConfig.token) {
            return;
        }
        var close$ = new rxjs_1.Subject();
        var state$ = new rxjs_1.BehaviorSubject({
            online: true,
            openPercent: 100,
        });
        var stateString$ = new rxjs_1.Subject();
        var device$ = nora_1.NoraService
            .getService(RED)
            .getConnection(noraConfig, this, stateString$)
            .pipe(operators_1.switchMap(function (connection) { return connection.addDevice(config.id, {
            type: 'blinds',
            name: config.devicename,
            roomHint: config.roomhint || undefined,
            state: state$.value,
        }); }), operators_1.publishReplay(1), operators_1.refCount(), operators_1.takeUntil(close$));
        device$.pipe(operators_1.switchMap(function (d) { return d.errors$; }), operators_1.takeUntil(close$)).subscribe(function (err) { return _this.warn(err); });
        rxjs_1.combineLatest([device$, state$])
            .pipe(operators_1.tap(function (_a) {
            var _ = _a[0], state = _a[1];
            return notifyState(state);
        }), operators_1.skip(1), operators_1.takeUntil(close$))
            .subscribe(function (_a) {
            var device = _a[0], state = _a[1];
            return device.updateState(state);
        });
        device$.pipe(operators_1.switchMap(function (d) { return d.state$; }), operators_1.takeUntil(close$)).subscribe(function (state) {
            notifyState(state);
            _this.send({
                payload: {
                    openPercent: adjustPercent(state.openPercent),
                },
                topic: config.topic
            });
        });
        this.on('input', function (msg) {
            if (config.passthru) {
                _this.send(msg);
            }
            util_1.updateState(msg === null || msg === void 0 ? void 0 : msg.payload, state$);
        });
        this.on('close', function () {
            close$.next();
            close$.complete();
        });
        function notifyState(state) {
            stateString$.next("(" + adjustPercent(state.openPercent) + "%)");
        }
        function adjustPercent(openPercent) {
            return config.invert ? 100 - openPercent : openPercent;
        }
    });
};
