// node-red-contrib-nora/build/nora-device.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var NoraDevice = /** @class */ (function () {
    function NoraDevice(id, config, connection) {
        var _this = this;
        this.id = id;
        this.config = config;
        this.connection = connection;
        this._stateChanged = new rxjs_1.Subject();
        this._activateScene = new rxjs_1.Subject();
        this.state$ = this._stateChanged.asObservable();
        this.activateScene$ = this._activateScene.asObservable();
        this.errors$ = this.connection.errors$.pipe(operators_1.filter(function (e) { return e.device === _this.id; }), operators_1.map(function (e) { return e.msg; }));
    }
    NoraDevice.prototype.updateState = function (partial) {
        var keys = Object.keys(partial);
        for (var _i = 0, keys_1 = keys; _i < keys_1.length; _i++) {
            var key = keys_1[_i];
            var newValue = partial[key];
            var oldValue = this.config.state[key];
            if (newValue !== oldValue) {
                this.config.state[key] = newValue;
            }
        }
        this.connection.sendDeviceUpdate(this.id, this.config.state);
    };
    NoraDevice.prototype.setState = function (newState) {
        this.config.state = newState;
        this._stateChanged.next(newState);
    };
    NoraDevice.prototype.activateScene = function (deactivate) {
        this._activateScene.next({ deactivate: deactivate });
    };
    return NoraDevice;
}());
exports.NoraDevice = NoraDevice;
